{data}[intel/2017a] GDAL 2.2.0 with support for HDF4 files.

jhein32 opened the pull request:
Fileset to create a GDAL 2.2.0 with support for HDF4 files.  Includes
an update for checksums in a few other GDAL configs.

boegel commented on easybuild/easyconfigs/h/HDF/HDF-4.2.13-intel-2017a-no-netcdf.eb:
```diff
@@ -0,0 +1,37 @@
+]
+
+configopts = "--with-szlib=$EBROOTSZIP --includedir=%(installdir)s/include/%(namelower)s"
+configopts += " --disable-netcdf --disable-fortran"
```
Hmm, these are not there in existing easyconfigs... Should they be?

What's the advantage of disabling netCDF support, or, vice versa, why don't we just include `netCDF` as a dependency and enable it? Same for Fortran, why don't we just enable it (which probably requires including `netCDF-Fortran`)?

jhein32 replied:
You raised that before.  In essence, this is a special version, tailored to the requirements of gdal.

boegel commented on easybuild/easyconfigs/h/HDF/HDF-4.2.13-intel-2017a-no-netcdf.eb:
```diff
@@ -0,0 +1,37 @@
+]
+dependencies = [
+    ('Szip', '2.1'),
+    ('libjpeg-turbo', '1.5.1'),
```
Bump to libjpeg-turbo 1.5.2? required easyconfig is already there...

jhein32 replied:
Done

boegel commented on easybuild/easyconfigs/h/HDF/HDF-4.2.13-intel-2017a-no-netcdf.eb:
```diff
@@ -0,0 +1,37 @@
+
+name = 'HDF'
+version = '4.2.13'
+versionsuffix = '-no-netcdf'
```
Since the other easyconfigs don't have this either, I wouldn't include this...

Instead, it would be better to enable netCDF (& Fortran) support?

jhein32 replied:
See above comment.  This is needed per gdal documentation.

boegel commented on easybuild/easyconfigs/g/GDAL/GDAL-2.2.0-intel-2017a-Python-2.7.13-HDF5-1.10.1-HDF-4.2.13.eb:
```diff
@@ -0,0 +1,53 @@
+version = '2.2.0'
+hdf5_ver = '1.10.1'
+hdf_ver = '4.2.13'
+versionsuffix = '-Python-%%(pyver)s-HDF5-%s-HDF-%s' % (hdf5_ver, hdf_ver)
```
This is becoming a bit of a mess...

I would include `Python` & `HDF5` (to keep consistency with existing easyconfigs), but I see no reason to also include `HDF`, especially since it's an uncommon dep?

jhein32 replied:
This is HDF4, we have a usergroup that uses HDF4 input files and am under the impression this is not that exotic, see below.  HDF4 is the purpose of this config.   

I wanted to make clear that this version supports HDF5 and HDF4.  

The GDAL website http://trac.osgeo.org/gdal/wiki/HDF devotes a fair amount of space how to build with support for both.  I assume this is of wider interest and thought I make it available.

You have a suggestion on how to clean up the name?

boegel replied:
Well, first of all, it's quite long, and the `-HDF5-<version>` part is only there because the `netCDF` used here happens to have that versionsuffix...

Including the `-HDF-<version>` in it makes sense to indicate this is a special variant of `GDAL`, but maybe it's overkill... Does also including `HDF` cripple the `GDAL` installation somehow(compared to a GDAL that doesn't, or does it just result in enhanced functionality?

If it's the latter, than I think including the `HDF` version in the versionsuffix is overkill (we don't list the version of any other dependencies either...).

boegel commented on easybuild/easyconfigs/g/GDAL/GDAL-2.2.0-intel-2017a-Python-2.7.13-HDF5-1.10.1-HDF-4.2.13.eb:
```diff
@@ -0,0 +1,53 @@
+source_urls = ['http://download.osgeo.org/gdal/%(version)s/']
+sources = [SOURCELOWER_TAR_XZ]
+checksums = ['0d4c326862e0f118e17418c042c2bcd037b25abd3fb198e1fc5d40b11a9fc8ea']
+patches = ['GDAL-%(version)s_fix-uchar-definition.patch']
```
Move up `patches` above `checksums`, also include a checksum for the patch?

You can let `eb --inject-checksums` do that for you, see http://easybuild.readthedocs.io/en/latest/Writing_easyconfig_files.html#adding-or-replacing-checksums-using-inject-checksums.

jhein32 replied:
Did a manual job to update the checksums

boegel commented on easybuild/easyconfigs/g/GDAL/GDAL-2.2.0-intel-2017a-Python-2.7.13-HDF5-1.10.1-HDF-4.2.13.eb:
```diff
@@ -0,0 +1,53 @@
+dependencies = [
+    ('Python', '2.7.13'),
+    ('netCDF', '4.4.1.1', '-HDF5-%s' % hdf5_ver),
+    ('HDF', '4.2.13', '-no-netcdf'),
```
It's a bit silly to have a `no-netcdf` version of HDF while we include `netCDF` as a dependency anyway?

jhein32 replied:
If you look at http://trac.osgeo.org/gdal/wiki/HDF, it is explicit that the netcdf inside HDF4 needs disabling for things to work.  

Sorry not my making.

boegel replied:
OK w.r.t. the `-no-netcdf`, since it covers the load. But you should use `hdf_ver` here rather than hardcoding `4.2.13` to make very sure that the HDF version mentioned in the `versionsuffix` matches the only that's actually used.

boegel commented on easybuild/easyconfigs/g/GDAL/GDAL-2.2.0-intel-2017a-Python-2.7.13-HDF5-1.10.1-HDF-4.2.13.eb:
```diff
@@ -0,0 +1,53 @@
+    ('SQLite', '3.17.0'),
+    ('libxml2', '2.9.4'),
+    ('libpng', '1.6.29'),
+    ('libjpeg-turbo', '1.5.1'),
```
Bump to libjpeg-turbo 1.5.2? required easyconfig is already there...

jhein32 replied:
Done

jhein32 commented:
@boegel  Hi Kenneth, could you please let me know whether you think this is interesting or to specialised.  If you think it is interesting, we can obviously rename this to something which has a better name, to make clear to end users that this is a special gdal with HDF4 support.

boegel commented on easybuild/easyconfigs/g/GDAL/GDAL-2.2.0-intel-2017a-Python-2.7.13-HDF5-1.10.1-HDF-4.2.13.eb:
```diff
@@ -0,0 +1,56 @@
+patches = ['GDAL-%(version)s_fix-uchar-definition.patch']
+checksums = [
+    '0d4c326862e0f118e17418c042c2bcd037b25abd3fb198e1fc5d40b11a9fc8ea',
+    '12daef0f16ebe438aeedc00b9eb2d35fdc5d39decffb567953459492a2a0fb5f',
```
When there's more than 1 checksums, we include a comment at the end of the line to mention the file (`--inject-checksums` does this automagically).

```python
checksums = [
    '0d4c326862e0f118e17418c042c2bcd037b25abd3fb198e1fc5d40b11a9fc8ea',  # gdal-2.2.0.tar.xz
    '12daef0f16ebe438aeedc00b9eb2d35fdc5d39decffb567953459492a2a0fb5f',  # GDAL-2.2.0_fix-uchar-definition.patch
]
```

If the line becomes too long (> 120 chars, like it does with the patch), put it above the checksum:

```python
checksums = [
    '0d4c326862e0f118e17418c042c2bcd037b25abd3fb198e1fc5d40b11a9fc8ea',  # gdal-2.2.0.tar.xz
    # GDAL-2.2.0_fix-uchar-definition.patch
    '12daef0f16ebe438aeedc00b9eb2d35fdc5d39decffb567953459492a2a0fb5f',
]
```

jhein32 replied:
@boegel On your first comment, I am not fully sure whether HDF cripples anything.  However the standard gdal's do not support HDF (only HDF5).  We need to make clear to users that this is a special version which does.

boegel commented on easybuild/easyconfigs/g/GDAL/GDAL-2.2.0-intel-2017a-Python-3.6.1.eb:
```diff
@@ -15,6 +15,7 @@ toolchainopts = {'usempi': True}
 source_urls = ['http://download.osgeo.org/gdal/%(version)s/']
 sources = [SOURCELOWER_TAR_XZ]
+checksums = ['0d4c326862e0f118e17418c042c2bcd037b25abd3fb198e1fc5d40b11a9fc8ea']
 patches = ['GDAL-%(version)s_fix-uchar-definition.patch']
```
checksum for patch is missing

jhein32 commented:
@boegel  Before merging, I need to understand a proj.4 dependency.  The users raised a query that his gdal usage failed because at runtime the proj.4 was not available.  If he loads the easybuild supplied PROJ/4.9.3 module everything works.  

I still struggle to see how gdal can build if the PROJ/4.9.3 is required.  This concerns him using a number of GDAL libs we have build in EasyBuild.  I asked him whether there is any user code involved that could cause the issue.

jhein32 commented:
@boegel I changed the filename and description to make to make it more compact and clearer what it is.  I added proj.4 to the dependencies, which seems a run time dependency.   Once would need to discuss whether this needs adding to other gdal versions.  I know that you are reluctant to do changes to released stuff.

jhein32 commented:
@boegel I hope this is now easier to work with.  One would need to discuss the proj.4 dependency (run time not build time), whether one wants this for other gdal as well.

boegel commented on easybuild/easyconfigs/g/GDAL/GDAL-2.2.0-intel-2017a-Python-2.7.13-HDF5-HDF.eb:
```diff
@@ -0,0 +1,59 @@
+name = 'GDAL'
+version = '2.2.0'
+hdf5_ver = '1.10.1'
+hdf_ver = '4.2.13'
```
@jhein32 `hdf5_ver` and `hdf_ver` are only used in one place now, so no need to use these local variables anymore?

jhein32 replied:
Only used in one place.  I was thinking the same but preferred it this way.  You feel strongly about it?

boegel commented:
@jhein32 If `PROJ` is a runtime dependency, then it should be listed as such. I'll look into a PR to update the existing `GDAL` easyconfigs to include it; updating existing easyconfigs is fine for bug fixes, and adding a missing dependency falls in the category.

boegel commented:
@jhein32 missing `PROJ` dependencies have been added via #5481

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 4 out of 4 (4 easyconfigs in this PR)
node2644.swalot.os - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/abe1e30c5d70446037cb9fa5ec0394f2 for a full test report.

boegel commented:
Test report by @boegel
**SUCCESS**
Build succeeded for 4 out of 4 (4 easyconfigs in this PR)
node2038.delcatty.os - Linux centos linux 7.4.1708, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/8e642fd3e6d8c1284bde30ea7a046b68 for a full test report.

boegel commented:
Going in, thanks @jhein32!
